modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4HttpRequestStreamIT.java

```diff
@@ -0,0 +1,124 @@
+        return false; // enable http
+    }
+
+    public void testBasicStream() throws IOException {
+        var totalBytes = 1024 * 1024;
+        var request = new Request("POST", RequestContentStreamPlugin.ROUTE);
+        request.setEntity(new ByteArrayEntity(randomByteArrayOfLength(totalBytes), ContentType.APPLICATION_JSON));
+
+        var respose = getRestClient().performRequest(request);
+        assertEquals(200, respose.getStatusLine().getStatusCode());
+        var gotTotalBytes = new BytesArray(respose.getEntity().getContent().readAllBytes()).utf8ToString();
+        assertEquals("" + totalBytes, gotTotalBytes);
+    }
+
+    public static class RequestContentStreamPlugin extends Plugin implements ActionPlugin {
+
+        static final String ROUTE = "/_test/request-stream/basic";
+        private static final Logger LOGGER = LogManager.getLogger("StreamRestHandler");
+
+        @Override
+        public Collection<RestHandler> getRestHandlers(
+            Settings settings,
+            NamedWriteableRegistry namedWriteableRegistry,
+            RestController restController,
+            ClusterSettings clusterSettings,
+            IndexScopedSettings indexScopedSettings,
+            SettingsFilter settingsFilter,
+            IndexNameExpressionResolver indexNameExpressionResolver,
+            Supplier<DiscoveryNodes> nodesInCluster,
+            Predicate<NodeFeature> clusterSupportsFeature
+        ) {
+            return List.of(new BaseRestHandler() {
+                @Override
+                public String getName() {
+                    return ROUTE;
+                }
+
+                @Override
+                public List<Route> routes() {
+                    return List.of(new Route(RestRequest.Method.POST, ROUTE));
+                }
+
+                @Override
+                protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) {
+                    return channel -> {
+
+                        // Example of handling streamed content
+
+                        var contentStream = request.contentStream();
+                        var totalReceivedBytes = new AtomicInteger();
+
+                        Consumer<HttpContent.Chunk> contentConsumer = (chunk) -> {
+                            LOGGER.info("got next chunk with size {}", chunk.bytes().length());
+                            totalReceivedBytes.addAndGet(chunk.bytes().length());
+                            if (chunk.isLast() == false) {
+                                contentStream.request(1024 * 10); // ask for 10kb, will round up to 16kb
+                            } else {
+                                channel.sendResponse(new RestResponse(RestStatus.OK, Integer.toString(totalReceivedBytes.get())));
+                            }
+                        };
+
+                        contentStream.setHandler(contentConsumer); // must setup handler before first request
```

Comment:
I wonder, rather than another layer of indirection here could we add a chunk-consuming method to `RestChannelConsumer`? Or possibly return something from `prepareRequest` which implements both `RestChannelConsumer` and a new `RequestBodyChunkConsumer` method to indicate we want to process a streaming body?

Reply:
I added  `RequestBodyChunkConsumer` 
```java
    public interface RequestBodyChunkConsumer extends RestChannelConsumer {
        void handleChunk(RestChannel channel, ReleasableBytesReference chunk, boolean isLast);
    }
```
and wiring for handler into body stream in `handleRequest`
```java
if (request.isStreamedContent()) {
  assert action instanceof RequestBodyChunkConsumer;
  var chunkConsumer = (RequestBodyChunkConsumer) action;
  request.contentStream().setHandler((chunk, isLast) -> chunkConsumer.handleChunk(channel, chunk, isLast));
}
```

implementation looks like this
```java
@Override
protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) {
    return new RequestBodyChunkConsumer() {

        int totalBytes = 0;

        @Override
        public void handleChunk(RestChannel channel, ReleasableBytesReference chunk, boolean isLast) {
            try (chunk) {
                totalBytes += chunk.length();
                if (isLast == false) {
                    request.contentStream().requestBytes(1024);
                } else {
                    channel.sendResponse(new RestResponse(RestStatus.OK, Integer.toString(totalBytes)));
                }
            }
        }

        @Override
        public void accept(RestChannel channel) throws Exception {
            request.contentStream().requestBytes(1024); // ask for first chunk
        }
    };
}

```